pkg/cache/v3/linear.go

```diff
@@ -312,90 +370,72 @@ func (cache *LinearCache) GetResources() map[string]types.Resource {
-	// of sending empty updates whenever an irrelevant resource changes.
-	stale := false
-	staleResources := []string{} // empty means all
-
-	// strip version prefix if it is present
-	var lastVersion uint64
-	var err error
-	if strings.HasPrefix(request.GetVersionInfo(), cache.versionPrefix) {
-		lastVersion, err = strconv.ParseUint(request.GetVersionInfo()[len(cache.versionPrefix):], 0, 64)
-	} else {
-		err = errors.New("mis-matched version prefix")
-	}
 
-	watch := ResponseWatch{Request: request, Response: value}
+	// A major difference between delta and sotw is the ability to not resend everything when connecting to a new control-plane
+	// In delta the request provides the version of the resources it does know, even if the request is wildcard or does request more resources
+	// In sotw the request only provides the global version of the control-plane, and there is no way for the control-plane to know if resources have
+	// been added since in the requested resources. In the context of generalized wildcard, even wildcard could be new, and taking the assumption
+	// that wildcard implies that the client already knows all resources at the given version is no longer true.
+	// We could optimize the reconnection case here if:
+	//  - we take the assumption that clients will not start requesting wildcard while providing a version. We could then ignore requests providing the resources.
+	//  - we use the version as some form of hash or resources known, and we can then consider it as a way to correctly verify whether all resources are unchanged.
+	// For now it is not done as:
+	//  - for the first case, while the protocol documentation does not explicitly mention the case, it does not mark it impossible and explicitly references unsubscribing from wildcard
+	//  - for the second one we could likely do it with little difficulty if need be
+	watch := ResponseWatch{Request: request, Response: value, subscription: sub}
 
 	cache.mu.Lock()
 	defer cache.mu.Unlock()
 
-	switch {
-	case err != nil:
-		stale = true
-		staleResources = request.GetResourceNames()
-		cache.log.Debugf("Watch is stale as version failed to parse %s", err.Error())
-	case len(request.GetResourceNames()) == 0:
-		stale = (lastVersion != cache.version)
-		if stale {
-			cache.log.Debugf("Watch is stale as cache version %d differs for wildcard watch %d", cache.version, lastVersion)
-		}
-	default:
-		for _, name := range request.GetResourceNames() {
-			// When a resource is removed, its version defaults 0 and it is not considered stale.
-			if lastVersion < cache.versionVector[name] {
-				stale = true
-				staleResources = append(staleResources, name)
-			}
-		}
-		if stale {
-			cache.log.Debugf("Watch is stale with stale resources %v", staleResources)
-		}
-	}
-	if stale {
-		cache.respond(watch, staleResources)
+	response := cache.computeSotwResponse(watch)
+	if response != nil {
+		watch.Response <- response
 		return nil, nil
 	}
```

Comment:
else this is an ack, right? might be worth a comment.

Reply:
Mentioned it above, there are other cases than ACK when no direct response comes in (e.g. adding a new resource in the subscription which currently does not exist in the cache)